tools/Dashboard/DevHome.Dashboard/ComSafeWidgetObjects/ComSafeWidgetDefinition.cs

```diff
@@ -0,0 +1,189 @@
+
+    private readonly ILogger _log = Log.ForContext("SourceContext", nameof(ComSafeWidgetDefinition));
+
+    private bool _hasValidProperties;
+    private const int MaxAttempts = 3;
+
+    public ComSafeWidgetDefinition(string widgetDefinitionId)
+    {
+        Id = widgetDefinitionId;
+    }
+
+    public async Task<bool> Populate()
+    {
+        await LazilyLoadOopWidgetDefinition();
+        return _hasValidProperties;
+    }
+
+    /// <summary>
+    /// Gets the theme resources from the widget. Tries multiple times in case of COM exceptions.
+    /// </summary>
+    /// <returns>The theme resources, or null in the case of failure.</returns>
+    public async Task<WidgetThemeResources> GetThemeResourceAsync(WidgetTheme theme)
+    {
+        var attempt = 0;
+        while (attempt++ < MaxAttempts)
+        {
+            try
+            {
+                await LazilyLoadOopWidgetDefinition();
+                return await Task.Run(() => _oopWidgetDefinition.GetThemeResource(theme));
+            }
+            catch (COMException ex) when (ex.HResult == RpcServerUnavailable || ex.HResult == RpcCallFailed)
+            {
+                _log.Warning(ex, $"Failed to operate on out-of-proc object with error code: 0x{ex.HResult:x}");
+                await GetNewOopWidgetDefinitionAsync();
+            }
+            catch (Exception ex)
+            {
+                _log.Error(ex, "Exception getting theme resources from widget:");
+                return null;
+            }
+        }
+
+        return null;
+    }
+
+    public async Task<WidgetCapability[]> GetWidgetCapabilitiesAsync()
+    {
+        var attempt = 0;
+        while (attempt++ < MaxAttempts)
+        {
+            try
+            {
+                await LazilyLoadOopWidgetDefinition();
+                return await Task.Run(() => _oopWidgetDefinition.GetWidgetCapabilities());
+            }
+            catch (COMException ex) when (ex.HResult == RpcServerUnavailable || ex.HResult == RpcCallFailed)
+            {
+                _log.Warning(ex, $"Failed to operate on out-of-proc object with error code: 0x{ex.HResult:x}");
+                await GetNewOopWidgetDefinitionAsync();
+            }
+            catch (Exception ex)
+            {
+                _log.Error(ex, "Exception getting widget capabilities from widget:");
+                return null;
+            }
+        }
+
+        return null;
+    }
+
+    private async Task GetNewOopWidgetDefinitionAsync()
+    {
+        _oopWidgetDefinition = null;
+        await LazilyLoadOopWidgetDefinition();
+    }
+
+    private async Task LazilyLoadOopWidgetDefinition()
+    {
+        var attempt = 0;
+        while (attempt++ < 3 && (_oopWidgetDefinition == null || _hasValidProperties == false))
+        {
+            try
+            {
+                _oopWidgetDefinition ??= await Application.Current.GetService<IWidgetHostingService>().GetWidgetDefinitionAsync(Id);

```

Comment:
Do we need to lock on this method in case two ore more threads attempted to race-load widgets at the same time?